README.md
# jQuery Quick Modal
Easily add customizable, CSS-driven modal windows to your website

See it in action here: <a href="http://kevinberonilla.com/demos/jquery-quick-modal">kevinberonilla.com/demos/jquery-quick-modal</a>

---

#### Dependencies
jQuery 1.7.0 or higher

---

#### Installation
Download the plugin via Bower or manually copy the CSS and JS files to your site, then include them in the `<head>` tag.
```
bower install jquery-quick-modal
```
```html
<head>
    <link rel="stylesheet" type="text/css" href="css/quick-modal.css">
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/1.11.3/jquery.min.js"></script>
    <script src="js/jquery.quick-modal.min.js"></script>
</head>
```

---

#### Create the Markup
```html
<a href="#" class="open-modal" data-modal-id="my-modal">Click here to open a modal window</a>
<div id="my-modal" class="qm-modal">
    <h1>Hello, world!</h1>
    <a href="#" class="qm-close-modal">Close window</a>
</div>
```
**`class="open-modal"`**  
The class for the element which will trigger a modal window on click; can be changed to a selector of your choice when initialized; default class is not styled by CSS

**`data-modal-id="my-modal"`**  
Specifies the ID of the modal that the link will be bound to; required (throws an error if not present)

**`class="qm-modal"`**   
The class for the modal window; has CSS styles that can be edited/overridden

**`class="qm-close-modal"`**   
The class for the element which will close an opened modal window on click; has CSS styles that can be edited/overridden; can be changed to a selector of your choice when initialized (loses CSS styles when done so)

---

#### Call the Script
```javascript
$(document).ready(function() {
    $('.open-modal').quickModal();
});
```
**`$('.open-modal')`**  
The jQuery selector object assigned to trigger modal windows on click

---

#### Play with Settings
```javascript
$(document).ready(function() {
    $('.open-modal').quickModal({ // Default values below
        animation: 'fade-zoom',
        speed: 250,
        timing: 'ease',
        closeModalSelector: '.qm-close-modal'
        enableEsc: true,
        enableClickAway: true,
        enableBodyScroll: false,
        appendBackgroundTo: 'body',
        prefix: 'qm',
        onOpen: function() {},
        onClose: function() {}
    });
});
```
**`animation`**  
The type of animation used for opening and closing modal windows
* `'fade'`
* `'fade-up'`
* `'fade-right'`
* `'fade-down'`
* `'fade-left'`
* `'fade-zoom'`
* `'fade-zoom-up'`
* `'fade-zoom-right'`
* `'fade-zoom-down'`
* `'fade-zoom-left'`

**`speed`**  
The length of time (in milliseconds) for the animation

**`timing`**   
The timing function used to animate the modal window
* `'linear'`
* `'ease'`
* `'ease-in'`
* `'ease-out'`
* `'ease-in-out'`
* `'cubic-bezier(n, n, n, n)'`

**`closeModalSelector`**   
The jQuery selector assigned to closing modal windows on click

**`enableEsc`**   
Boolean for determining whether or not pressing the Esc key will close modals

**`enableClickAway`**   
Boolean for determining whether or not clicking the modal background will close modals

**`enableBodyScroll`**   
Boolean for determining whether or not the body behind an open modal window is scrollable

**`appendBackgroundTo`**   
The jQuery selector to which the modal background will be appended

**`prefix`**   
String representing the prefix used for all classes and ids (only needed for builds with custom prefixed CSS; does not directly affect `closeModalSelector`)

**`onOpen`**   
The callback that fires when the modal is opened

**`onClose`**   
The callback that fires when the modal is closed

---

#### Trigger Modal Windows Manually
```javascript
$(document).ready(function() {
    $('#my-modal').quickModal('open');
    
    $('#my-modal').quickModal('open', {
        animation: 'fade-zoom-down',
        speed: 500,
        timing: 'ease-in-out',
        closeModalSelector: '.close-button'
    });
    
    $('#my-modal').quickModal('close');
    
    $('#my-modal').quickModal('close', {
        animation: 'fade-left',
        speed: 750,
        timing: 'linear'
    });
    
    $('#modal-link').quickModal('trigger');
    
    $('#modal-link').quickModal('trigger', {
        animation: 'fade',
        timing: 'ease-out'
    });
});
```
**`$('#my-modal')`**  
The jQuery selector object for the modal window to be triggered

**`$('#modal-link')`**  
The jQuery selector object for the element that will trigger its modal window

**`.quickModal('open')`**  
Open the selected modal window with default settings

**`.quickModal('open', {...})`**  
Open the selected modal window with the specified settings

**`.quickModal('close')`**  
Close the selected modal window with previously applied settings

**`.quickModal('close', {...})`**  
Close the selected modal window with the specified settings

**`.quickModal('trigger')`**  
Open the target of the selected element with default settings

**`.quickModal('trigger', {...})`**  
Open the target of the selected element with the specified settings

---

#### Events
```javascript
$(document).ready(function() {
    $('#my-modal').on('modalopen.qm', myAwesomeFunction);
    
    $('#my-modal').on('modalclose', function() {
        alert('The modal window has been closed!');
        myAwesomeFunction();
    });
});
```
**`'modalopen.qm'`**  
The event fired immediately before a modal window is opened

**`'modalclose.qm'`**  
The event fired immediately after a modal window is closed

---

#### Call the Script with AJAX Compatibility
```javascript
$(document).ready(function() {
    $('body').on('click', '.open-modal', function() {
        $(this).quickModal('trigger');
    });
});
```
This alternative way of initializing the plugin will ensure that elements that match the selector will always open its associated modal window, regardless of whether or not they were present on page load.

---

#### Using Custom Prefixes

Included in the package are Sass source files (with a variable for managing custom prefixed classes and ids) and Gulp build scripts. Be sure to use the `prefix` option in the plugin to support your custom prefix.


CHANGELOG.md
# jQuery Quick Modal

#### v2.1.0 | August 28, 2017
* Fixed an issue where modal backdrops were not animating properly when closing a modal
* Added the `prefix` option to support builds with custom prefixes
* Added Sass source files and Gulp tasks to support builds with custom prefixes
* The plugin now runs in strict mode

---

#### v2.0.0 | August 22, 2017
* Updated all HTML classes and ids to use the `qm-` prefix
* Updated all events to use the `.qm` namespace
* Updated versioning to semantic style
* Removed CSS vendor prefixes (no longer needed)
* Demoted console logs for incorrect configuration to `warn` level (was `error` level)

---

#### v1.08 | October 5, 2015
* Fixed an issue where the close modal bindings still persisted when a specific combination of actions were taken

---

#### v1.07 | October 5, 2015
* Added the `onOpen` and `onClose` options
* Fixed an issue where the close modal bindings persisted after unbinding
* Code cleanup for maintainability

---

#### v1.06 | September 30, 2015
* Added the `appendBackgroundTo` option
* Fixed an issue where two modal backgrounds could be appended through a specific combination of bindings

---

#### v1.05 | September 16, 2015
* Code cleanup for maintainability

---

#### v1.04 | August 3, 2015
* Added the `enableBodyScroll` option

---

#### v1.03 | July 30, 2015
* Fixed an issue where the unbinding of the Esc key prevented modals to open on Firefox

---

#### v1.02 | July 22, 2015
* Added the `enableEsc` and `enableClickAway` options
* Added the `trigger` method
* Fixed an issue where modals taller than the viewport were not scrolling correctly

---

#### v1.01 | July 14, 2015
* Added method functionality
* Fixed an issue where the close modal selector was not binding correctly
* Added a min-height CSS property on the `.modal` class (fixes the appearance of empty modals)

---

#### v1.00 | July 9, 2015
* So it begins...

gulpfile.js
var gulp = require('gulp');
var cssnano = require('gulp-cssnano');
var uglify = require('gulp-uglify');
var insert = require('gulp-insert');
var rename = require('gulp-rename');
var runSequence = require('run-sequence');
var sass = require('gulp-sass');
var autoprefixer = require('gulp-autoprefixer');
var header = 
`/* --------------------------------------------------
jQuery Quick Modal v2.1.0

By Kevin Beronilla
http://www.kevinberonilla.com

Fork on GitHub
https://github.com/kevinberonilla/jquery-quick-modal

Free to use under the MIT license
http://www.opensource.org/licenses/mit-license.php
-------------------------------------------------- */
`;

gulp.task('compile:sass', () => {
    return gulp.src(['./scss/*.scss'])
        .pipe(sass({ outputStyle: 'expanded' }).on('error', sass.logError))
        .pipe(autoprefixer())
        .pipe(gulp.dest('./css'));
});

gulp.task('minify:css', () => {
    return gulp.src(['./css/*.css', '!./css/*.min.css'])
        .pipe(cssnano())
        .pipe(rename({ suffix: '.min' }))
        .pipe(gulp.dest('./css'));
});

gulp.task('minify:js', () => {
    return gulp.src(['./js/*.js', '!./js/*.min.js'])
        .pipe(uglify())
        .pipe(rename({ suffix: '.min' }))
        .pipe(gulp.dest('./js'));
});

gulp.task('insert:comments:css', () => {
    return gulp.src(['./css/*.min.css'])
        .pipe(insert.prepend(header))
        .pipe(gulp.dest('./css'));
});

gulp.task('insert:comments:js', () => {
    return gulp.src(['./js/*.min.js'])
        .pipe(insert.prepend(header))
        .pipe(gulp.dest('./js'));
});

gulp.task('watch', () => {
    gulp.watch(['./scss/*.scss'], ['compile:sass']);
    gulp.watch(['./css/*.css', '!./css/*.min.css'], ['minify:css']);
    gulp.watch(['./js/*.js', '!./js/*.min.js'], ['minify:js']);
    gulp.watch(['./css/*.css'], ['insert:comments:css']);
    gulp.watch(['./js/*.min.js'], ['insert:comments:js']);
});

gulp.task('default', () => {
    return runSequence('compile:sass', 'minify:css', 'minify:js', 'insert:comments:css', 'insert:comments:js', 'watch');
});

js/jquery.quickmodal.js
/* --------------------------------------------------
jQuery Quick Modal v1.00

By Kevin Beronilla
http://www.kevinberonilla.com

Fork on GitHub
https://github.com/kevinberonilla/jquery-quickmodal

Free to use under the MIT license
http://www.opensource.org/licenses/mit-license.php
-------------------------------------------------- */
(function($) { // Protect the $ alias (IIF)
    $.fn.quickmodal = function(args) {
        if (!$('#modal-background').length) $('body').append('<div id="modal-background"></div>'); // Append background; do not append if re-initialized or background already exists
        
        if (args !== null && typeof args === 'string') { // Calling method
            var bodyTag = $('body'),
                closeModalLink = $('.close-modal'),
                modalBackground = $('#modal-background');
            
            switch (args) {
                case 'open':
                    bodyTag.addClass('disable-scroll');
                    modalBackground.show();
                    this.show();
                    setTimeout(function() { // Ensure elements are displayed before adding classes
                        modalBackground.addClass('visible');
                        $(this).addClass('visible');
                    }, 25);
                    break;
                    
                case 'close':
                    bodyTag.removeClass('disable-scroll');
                    modalBackground.removeClass('visible');
                    this.removeClass('visible');
                    setTimeout(function() {
                        modalBackground.hide();
                        this.hide();
                    }, 250);
                    break;
            }
        } else { // Initializing plugin with options
            var settings = $.extend({ // Extend the default settings established below
                    modalWindowClass: '.modal',
                    closeModalClass: '.close-modal'
                    
                    // ADD CALLBACK OPTIONS
                
                
                }, args),
                bodyTag = $('body'),
                openModalLink = this,
                modalWindow = $(settings.modalWindowClass),
                closeModalLink = $(settings.closeModalClass),
                modalBackground = $('#modal-background');
            
            function closeModal() {
                bodyTag.removeClass('disable-scroll');
                modalBackground.removeClass('visible');
                modalWindow.removeClass('visible');
                setTimeout(function() {
                    modalBackground.hide();
                    modalWindow.hide();
                }, 250);
            }
            
            openModalLink.click(function(e) {
                e.preventDefault();
                
                var modalId = $(this).data('modal-id'),
                    targetModal = $('#' + modalId);
                
                if (modalId === undefined) console.error('No "data-modal-id" attribute is set.');
                
                bodyTag.addClass('disable-scroll');
                modalBackground.show();
                targetModal.show();
                setTimeout(function() { // Ensure elements are displayed before adding classes
                    modalBackground.addClass('visible');
                    $('#' + modalId).addClass('visible');
                }, 25);
            });
            
            closeModalLink.click(function(e) {
                e.preventDefault();
                closeModal();
            });
            
            $(document).keyup(function(e) {
                if (e.keyCode == 27) closeModal(); // Esc
            });
            
            modalBackground.click(closeModal);
        }
        
        return this; // Return the object to enable chaining
    }
}(jQuery));

js/jquery.quick-modal.js
/* --------------------------------------------------
jQuery Quick Modal v2.0.0

By Kevin Beronilla
http://www.kevinberonilla.com

Fork on GitHub
https://github.com/kevinberonilla/jquery-quick-modal

Free to use under the MIT license
http://www.opensource.org/licenses/mit-license.php
-------------------------------------------------- */
(function($) { // Protect the $ alias (IIF)
    'use strict';
    
    $.fn.setSpeed = function(speed) {
        return this.css({
            '-webkit-transition-duration': speed + 'ms',
            'transition-duration': speed + 'ms'
        });
    }
    
    $.fn.setTiming = function(timing) {
        return this.css({
            '-webkit-transition-timing-function': timing,
            'transition-timing-function': timing
        });
    }
    
    function checkSettings(modalObj, backgroundObj, settings) {
        modalObj.setSpeed(null)
            .setTiming(null);
        backgroundObj.setSpeed(null)
            .setTiming(null);
        
        if (settings.speed != 250 || settings.timing != 'ease') { // Set CSS if settings not equal to default
            modalObj.setSpeed(settings.speed)
                .setTiming(settings.timing);
            backgroundObj.setSpeed(settings.speed)
                .setTiming(settings.timing);
        }
    }
    
    $.fn.quickModal = function(args, options) {
        if (args !== null && typeof args === 'string') { // If calling a method
            var settings = $.extend({ // Extend the default settings established below
                    animation: 'fade-zoom',
                    speed: 250,
                    timing: 'ease',
                    closeModalSelector: '.qm-close-modal',
                    enableEsc: true,
                    enableClickAway: true,
                    enableBodyScroll: false,
                    appendBackgroundTo: 'body',
                    prefix: 'qm',
                    onOpen: function() {},
                    onClose: function() {}
                }, options),
                bodyTag = $('body'),
                closeModalLink = $(settings.closeModalSelector),
                modal = $('.' + settings.prefix + '-modal'),
                self = this;
            
            function keyUpCheck(e) {
                if (e.keyCode == 27 && modal.is(':visible') && settings.enableEsc) closeModal(); // Esc key events based on options
            }
            
            function closeModal() {
                self.quickModal('close', settings);
                bodyTag.unbind('keyup', keyUpCheck);
                closeModalLink.unbind('click');
                $('#' + settings.prefix + '-modal-background').unbind('click');
            }
            
            if (!$('#modal-background').length) $(settings.appendBackgroundTo).append('<div id="' + settings.prefix + '-modal-background"></div>'); // Append background; do not append if background already exists
            
            checkSettings(self, $('#' + settings.prefix + '-modal-background'), settings);
            modal.removeClass()
                .addClass(settings.prefix + '-modal')
                .addClass(settings.prefix + '-animation-' + settings.animation);
            
            switch (args) {
                case 'open':
                    if (!settings.enableBodyScroll) bodyTag.addClass(settings.prefix + '-disable-scroll');
                    
                    modal.hide(); // Hide any currently visible modals
                    self.show();
                    bodyTag.keyup(keyUpCheck);
                    $('#' + settings.prefix + '-modal-background').show();
                    
                    setTimeout(function() { // Ensure elements are displayed before adding classes
                        if (settings.enableClickAway) $('#' + settings.prefix + '-modal-background').click(closeModal);
                        $('#' + settings.prefix + '-modal-background').addClass(settings.prefix + '-visible');
                        self.addClass(settings.prefix + '-visible');
                        self.trigger('modalopen.qm'); // Trigger custom 'open' event
                        settings.onOpen.call(); // Open callback
                    }, 25);
                    
                    closeModalLink.click(function(e) { // Bind events based on options
                        e.preventDefault();
                        closeModal();
                    });
                    
                    break;
                    
                case 'close':
                    bodyTag.removeClass(settings.prefix + '-disable-scroll');
                    $('#' + settings.prefix + '-modal-background').removeClass(settings.prefix + '-visible');
                    self.removeClass(settings.prefix + '-visible');
                    settings.onClose.call(); // Close callback
                    
                    setTimeout(function() {
                        $('#' + settings.prefix + '-modal-background').hide();
                        self.hide();
                        self.trigger('modalclose.qm'); // Trigger custom 'close' event
                    }, settings.speed);
                    
                    break;
                    
                case 'trigger':
                    var modalId = self.data('modal-id'),
                    targetModal = $('#' + modalId);
                    
                    targetModal.quickModal('open', settings);
                    
                    break;
                    
                default:
                    console.warn('The method you entered does not exist.');
            }
        } else { // If initializing plugin with options
            var self = this;
            
            self.click(function(e) {
                e.preventDefault();
                
                var modalId = $(this).data('modal-id'),
                    targetModal = $('#' + modalId);
                
                if (modalId === undefined) console.warn('No "data-modal-id" attribute has been set.');
                else targetModal.quickModal('open', args);
            });
        }
        
        return this; // Return the object to enable chaining
    }
}(jQuery));